FirebaseSegmentation/Sources/SEGContentManager.m

```diff
@@ -0,0 +1,119 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "SEGContentManager.h"
+
+#import <FirebaseCore/FIRAppInternal.h>
+#import <FirebaseInstanceID/FIRInstanceID.h>
+#import "FIRSegmentation.h"
+#import "SEGDatabaseManager.h"
+#import "SEGNetworkManager.h"
+#import "SEGSegmentationConstants.h"
+
+@interface SEGContentManager () {
+  NSMutableDictionary<NSString *, id> *_associationData;
+  NSString *_instanceIdentifier;
+  NSString *_instanceIdentifierToken;
+  SEGDatabaseManager *_databaseManager;
+  SEGNetworkManager *_networkManager;
+}
+@end
+
+@implementation SEGContentManager
+
++ (instancetype)sharedInstanceWithFIROptions:(FIROptions *)options {
+  static dispatch_once_t onceToken;
+  static SEGContentManager *sharedInstance;
+  dispatch_once(&onceToken, ^{
+    sharedInstance = [[SEGContentManager alloc]
+        initWithDatabaseManager:[SEGDatabaseManager sharedInstance]
+                 networkManager:[[SEGNetworkManager alloc] initWithFIROptions:options]];
+  });
+  return sharedInstance;
+}
+
+- (instancetype)initWithDatabaseManager:databaseManager networkManager:networkManager {
+  self = [super init];
+  if (self) {
+    // Initialize the database manager.
+    _databaseManager = databaseManager;
+
+    // Initialize the network manager.
+    _networkManager = networkManager;
+
+    NSAssert(_databaseManager != nil, @"Segmentation database could not be initialized");
+
+    // Load all data from the database.
+    [_databaseManager createOrOpenDatabaseWithCompletion:^(BOOL success, NSDictionary *result) {
+      self->_associationData = [result mutableCopy];
+    }];
+    // TODO(dmandar) subscribe to FIS notifications once integrated.
+  }
+  return self;
+}
+
+// TODO(dmandar) IID only supports default instance. Modify for FIS.
+- (FIRInstanceID *)instanceIDForApp:(NSString *)firebaseApp {
+  return [FIRInstanceID instanceID];
+}
+
+- (void)associateCustomInstallationIdentiferNamed:(NSString *)customInstallationID
+                                      firebaseApp:(NSString *)firebaseApp
+                                       completion:(SEGRequestCompletion)completionHandler {
+  // Get the latest instance identifier
+  if (![self instanceIDForApp:firebaseApp]) {
+    completionHandler(NO, @{@"ErrorDescription" : @"InstanceID SDK not available"});
+  }
+  __weak SEGContentManager *weakSelf = self;
+  [[FIRInstanceID instanceID] instanceIDWithHandler:^(FIRInstanceIDResult *_Nullable result,
+                                                      NSError *_Nullable error) {
```

Comment:
It looks like we don't actually check if there was an error here or not - this means it could wipe out all the existing information on a fetch failure. Is that expected?

Reply:
good catch. done.